File: apps/trace-propagation/http/node/handlers.js
const tracer = require('dd-trace').init();
const { sendDistributionMetric } = require('datadog-lambda-js');

const serverUrls = process.env.SERVER_URLS.split(',');
const runtime = process.env.AWS_EXECUTION_ENV.replace('AWS_Lambda_', '');

exports.client = async function(event, context) {
  data = JSON.stringify({
    runtime: runtime,
    trace_id: currentTraceId(),
  });
  for (url of serverUrls) {
    console.log(`calling ${url} with data ${data}`);
    await fetch(url, {body: data});
  }
  return {'statusCode': 200, 'body': 'ok'};
}

exports.server = async function(event, context) {
  const traceId = currentTraceId();
  const { body } = event;
  console.log(`received http data ${body}`);
  payload = JSON.parse(body);
  sendDistributionMetric(
    'trace_context.propagated.http', 1,
    `server_runtime:${runtime}`,
    `client_runtime:${payload.runtime}`,
    traceId == payload.trace_id ? 'success:true' : 'success:false',
    'transport:http',
  );
  return {'statusCode': 200, 'body': 'ok'};
}

function currentTraceId() {
  const ctx = tracer.scope().active()?.context();
  console.log(`found trace context: traceId=${ctx.toTraceId()} spanId=${ctx.toSpanId()}`);
  return ctx.toTraceId();
}

File: apps/trace-propagation/http/python/handlers.py
import base64
import json
import requests
import os

from datadog_lambda.metric import lambda_metric
from ddtrace import tracer

server_urls = os.environ.get('SERVER_URLS', '').split(',')
runtime = os.environ.get('AWS_EXECUTION_ENV', '').replace('AWS_Lambda_', '')

def client(event, context):
    data = json.dumps({
            'runtime': runtime,
            'trace_id': _current_trace_id(),
    })
    for url in server_urls:
        print(f'calling {url} with data {data}')
        requests.get(url, data=data)
    return {'statusCode': 200, 'body': 'ok'}

def server(event, context):
    trace_id = _current_trace_id()
    body = event.get('body')
    print(f'received http event body {body}')
    payload = json.loads(base64.b64decode(body))
    print(f'received http data {payload}')
    lambda_metric('trace_context.propagated.http', 1, tags=[
            f'client_runtime:{runtime}',
            f'server_runtime:{payload.get("runtime")}',
            f'success:{trace_id == payload.get("trace_id")}',
            f'transport:http',
    ])
    return {'statusCode': 200, 'body': 'ok'}

def _current_trace_id():
    ctx = tracer.current_trace_context()
    assert ctx, 'no trace context found!'
    print(f'found trace context {ctx}')
    return str(ctx.trace_id)
